### PR TITLE
drivers: sensor: bmi270: Add support for motion, DRDY triggers

### PR DESCRIPTION
**Pull request by tautologyclub**
Trigger support for BMI270.

Note: To use these features, you can't use the "max fifo" blob previously included in the driver because that's feature stripped (in favor of max fifo size). You need to ship the new blob. That's accomplished by appending a compat string in your dts node:

```
	bmi270@69 {
		compatible = "bosch,bmi270", "bosch,bmi270-base";
		reg = <0x69>;
		irq-gpios = <&gpio0 2 (GPIO_ACTIVE_LOW)>,
			    <&gpio1 15 (GPIO_ACTIVE_LOW)>;
	};
```

Clearly a sample would be good to demonstrate this. Patches welcome ;)

### CODE REVIEW

File: drivers/sensor/bmi270/bmi270.c [hunk 1]
```diff
@@ -715,33 +773,69 @@ static int bmi270_init(const struct device *dev)
...
+	.config_file_len = sizeof(bmi270_config_file_base),
+	.anymo_1 = &(struct bmi270_feature_reg){ .page = 1, .addr = 0x3C },
+	.anymo_2 = &(struct bmi270_feature_reg){ .page = 1, .addr = 0x3E },
+};
+
+static const struct bmi270_feature_config bmi270_feature_legacy = {
+	.name = "legacy",
+	.config_file = bmi270_config_file_legacy,
+	.config_file_len = sizeof(bmi270_config_file_legacy),
+};
```
Comment (MaureenHelm) [inline drivers/sensor/bmi270/bmi270.c:800]: What is the difference between legacy and base? Can you please explain the changes to the config blob?

File: drivers/sensor/bmi270/bmi270.c [hunk 2]
```diff
@@ -552,14 +550,57 @@ static int bmi270_channel_get(const struct device *dev, enum sensor_channel chan
...
 {
 	int ret = -ENOTSUP;
 
-	if ((chan == SENSOR_CHAN_ACCEL_X) || (chan == SENSOR_CHAN_ACCEL_Y)
-	    || (chan == SENSOR_CHAN_ACCEL_Z)
-	    || (chan == SENSOR_CHAN_ACCEL_XYZ)) {
+	if ((chan == SENSOR_CHAN_ACCEL_X) ||
+	    (chan == SENSOR_CHAN_ACCEL_Y) ||
+	    (chan == SENSOR_CHAN_ACCEL_Z) ||
+	    (chan == SENSOR_CHAN_ACCEL_XYZ)) {
```
Comment (MaureenHelm) [inline drivers/sensor/bmi270/bmi270.c:603]: This change isn't related to adding triggers and should be made in a separate commit

File: drivers/sensor/bmi270/bmi270.c [hunk 3]
```diff
@@ -570,12 +611,19 @@ static int bmi270_attr_set(const struct device *dev, enum sensor_channel chan,
...
 		default:
 			ret = -ENOTSUP;
 		}
-	} else if ((chan == SENSOR_CHAN_GYRO_X) || (chan == SENSOR_CHAN_GYRO_Y)
-		   || (chan == SENSOR_CHAN_GYRO_Z)
-		   || (chan == SENSOR_CHAN_GYRO_XYZ)) {
+	} else if ((chan == SENSOR_CHAN_GYRO_X) ||
+		   (chan == SENSOR_CHAN_GYRO_Y) ||
+		   (chan == SENSOR_CHAN_GYRO_Z) ||
+		   (chan == SENSOR_CHAN_GYRO_XYZ)) {
```
Comment (MaureenHelm) [inline drivers/sensor/bmi270/bmi270.c:626]: This change isn't related to adding triggers

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please address the failed compliance checks

File: drivers/sensor/bmi270/bmi270.c (hunk 1 above)
Comment (tautologyclub) [reply drivers/sensor/bmi270/bmi270.c:800]: The BMI270 is basically firmware flashed at each boot. The blobs enable entirely different feature sets. To get triggers, you can't use max_fifo. This driver doesn't implement any features from the `legacy` blob tho so I guess that part could be removed...

**Comment by tautologyclub** [comment]
Compliance issues fixed

File: drivers/sensor/bmi270/bmi270_trigger.c [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+			       struct gpio_callback *pin_cb,
+			       gpio_callback_handler_t handler)
+{
+	int ret;
+
+	if (!pin->port) {
+		return 0;
+	}
+
+	if (!device_is_ready(pin->port)) {
```
Comment (bbilas) [inline drivers/sensor/bmi270/bmi270_trigger.c:141]:
```suggestion
	if (!gpio_is_ready_dt(pin)) {
```

File: drivers/sensor/bmi270/bmi270.c (hunk 2 above)
Comment (tautologyclub) [reply drivers/sensor/bmi270/bmi270.c:603]: I have changed this and verified that I've pushed it, but it seems as if something is configured to ignore "whitespace only" changes and I don't think it's locally. Would you mind either a) letting it slide, or b) removing the added newline yourself?

File: drivers/sensor/bmi270/bmi270_trigger.c (hunk 4 above)
Comment (tautologyclub) [reply drivers/sensor/bmi270/bmi270_trigger.c]: ok done

File: drivers/sensor/bmi270/bmi270.c (hunk 2 above)
Comment (MaureenHelm) [reply drivers/sensor/bmi270/bmi270.c:603]: Please check again. I don't object to the change, but it needs to be in a separate commit so it doesn't add noise to functional changes.

File: drivers/sensor/bmi270/bmi270_trigger.c [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+	switch (trig->type) {
+	case SENSOR_TRIG_MOTION:
+		if (!cfg->int1.port) {
+			return -ENOTSUP;
+		}
+
+		k_mutex_lock(&data->trigger_mutex, K_FOREVER);
+		data->motion_handler = handler;
+		data->motion_trigger = *trig;
```
Comment (MaureenHelm) [inline drivers/sensor/bmi270/bmi270_trigger.c:289]: Store the trigger pointer instead of making a copy; this will enable the handler to use `CONTAINER_OF`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> Please address the failed compliance checks

This still needs to be done. You can ignore the long line warnings in the firmware blob, but there are additional warnings that need to be addressed.

File: drivers/sensor/bmi270/bmi270.c (hunk 1 above)
Comment (tautologyclub) [reply drivers/sensor/bmi270/bmi270.c]: But the new firmware blob breaks compatibility. It's essentially becomes a different device, albeit using the same silicon. To me this seems like a pretty good place to utilize the compat string. I think making a whole new driver would be more close to correct than introducing a property.

**Comment by tautologyclub** [comment]
@MaureenHelm Something went wrong last time, my fixes didn't get pushed :O should be okay now :)

**Comment by tautologyclub** [comment]
Was there anything I missed?

**Review by MaureenHelm — APPROVED** [review]

**Review by teburd — APPROVED** [review]
